userspace/async/async_metadata_source.tpp

```diff
@@ -0,0 +1,286 @@
+/*
+Copyright (C) 2018 Sysdig, Inc.
+
+This file is part of sysdig.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+*/
+#include "logger.h"
+
+#include <assert.h>
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <list>
+#include <string>
+
+namespace sysdig
+{
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::async_metadata_source(
+		const uint64_t max_wait_ms,
+		const uint64_t ttl_ms):
+	m_max_wait_ms(max_wait_ms),
+	m_ttl_ms(ttl_ms),
+	m_thread(),
+	m_running(false),
+	m_terminate(false),
+	m_mutex(),
+	m_queue_not_empty_condition(),
+	m_metadata_map()
+{ }
+
+template<typename key_type, typename metadata_type>
+async_metadata_source<key_type, metadata_type>::~async_metadata_source()
+{
+	try
```

Comment:
It isn't throwing any user-specified exceptions, that's true, but it could be calling something that transitively throws an exception.  If an exception is thrown from the destructor, it'll crash the program (in C++11), so I'm being defensive against that.

Reply:
in c++ do you not need to rethrow the exception at each layer of the stacK?

in either case, AFAIK we don't catch exceptions in most places, so I'm not sure without specific reasons we should be catching exceptions without a good understanding of what they are and why they might occur.....pretty much why do we have a requirement that THIS particular part of code need to be safer than the entire rest of the codebas?

Reply:
No.  Destructors must not throw exceptions.

Far from authoritative, but:
https://akrzemi1.wordpress.com/2013/08/20/noexcept-destructors/

I'm not in favor of using the existing code base as the high-bar for new code.

If we want the agent to crash if an exception is thrown here, I can remove the try/catch.  If we do not, then I'm in favor of leaving it.